src/nvim/os/time.c

```diff
@@ -46,31 +47,70 @@ void os_delay(uint64_t milliseconds, bool ignoreinput)
     }
     LOOP_PROCESS_EVENTS_UNTIL(&main_loop, NULL, (int)milliseconds, got_int);
   } else {
-    os_microdelay(milliseconds * 1000);
+    os_microdelay(milliseconds * 1000u, ignoreinput);
   }
 }
 
-/// Sleeps for a certain amount of microseconds
+/// Sleeps for a certain amount of microseconds.
 ///
-/// @param microseconds Number of microseconds to sleep
-void os_microdelay(uint64_t microseconds)
+/// @param microseconds Number of microseconds to sleep.
+/// @param ignoreinput If true, ignore all input (including SIGINT/CTRL-C).
+///                    If false, waiting is aborted on any input.
+void os_microdelay(uint64_t microseconds, bool ignoreinput)
 {
-  uint64_t elapsed = 0;
-  uint64_t ns = microseconds * 1000;  // convert to nanoseconds
+  uint64_t elapsed = 0u;
   uint64_t base = uv_hrtime();
+  const uint64_t input_check_interval = 100000000u;  // 100ms
+
+  // Convert microseconds to nanoseconds. If uint64_t would overflow, set
+  // nanoseconds to UINT64_MAX.
+  const uint64_t nanoseconds = (microseconds < UINT64_MAX/1000u)
+                               ? microseconds * 1000u
+                               : UINT64_MAX;
 
   uv_mutex_lock(&delay_mutex);
 
-  while (elapsed < ns) {
-    if (uv_cond_timedwait(&delay_cond, &delay_mutex, ns - elapsed)
-        == UV_ETIMEDOUT)
+  // If the input is ignored, we simply wait the full delay. If not, we
+  // check every 'input_check_delay' milliseconds for input and break the
+  // waiting loop if input is available.
+  while (elapsed < nanoseconds) {
+    // We split the waiting period in 'input_check_interval' long intervals in
+    // case of 'ignoreinput == true' or wait the whole 'nanoseconds' if
+    // 'ignoreinput == false'.
+    const uint64_t nanoseconds_delta = (ignoreinput)
+                                       ? nanoseconds - elapsed
+                                       : MIN(nanoseconds - elapsed,
+                                             input_check_interval);
+
+    // Main waiting work happens here.
+    if (UV_ETIMEDOUT != uv_cond_timedwait(&delay_cond, &delay_mutex,
+                                          nanoseconds_delta)) {
+      // In case of an error in uv_cond_timedwait() we continue quietly in
+      // release mode and terminate with an error message in debug mode.
```

Comment:
So, let’s look at `man pthread_cond_timedwait`:

> Upon successful completion, a value of zero shall be returned; otherwise, an error number shall be returned to indicate the error.